objects/src/accounts/vault.rs

```diff
@@ -59,11 +65,161 @@ impl AccountVault {
         if !matches!(faucet_id.account_type(), AccountType::FungibleFaucet) {
             return Err(AccountError::not_a_fungible_faucet_id(faucet_id));
         }
-        todo!()
+
+        // if the tree value is [0, 0, 0, 0], the asset is not stored in the vault
+        match self.asset_tree.get_value([ZERO, ZERO, ZERO, faucet_id.into()].into()) {
+            asset if asset == [ZERO, ZERO, ZERO, ZERO] => Ok(0),
+            asset => Ok(FungibleAsset::try_from(asset)
+                .expect("tree only contains valid assets")
+                .amount()),
+        }
+    }
+
+    /// Returns an iterator over the assets stored in the vault.
+    pub fn assets(&self) -> impl Iterator<Item = Asset> + '_ {
+        self.asset_tree
+            .bottom_leaves()
+            .flat_map(|(_, values)| {
+                values
+                    .iter()
+                    .map(|value| Asset::try_from(value.1).expect("tree only contains valid assets"))
+                    .collect::<Vec<_>>()
+            })
+            .chain(self.asset_tree.upper_leaves().map(|(_, _, value)| {
+                Asset::try_from(value).expect("tree only contains valid assets")
+            }))
+    }
+
+    /// Returns a reference to the asset TSMT.
+    pub fn asset_tree(&self) -> &TieredSmt {
+        &self.asset_tree
+    }
```

Comment:
Question: why do we need to expose this publicly? If possible, we should try to keep this as an internal implementation detail.

Reply:
We only expose this to support population of the advice provider for the `Account` object. Lets go with your proposal of implementing `ToAdviceInputs` on the `AccountVault` and we can remove this.